suave/builder/session_manager.go

```diff
@@ -0,0 +1,138 @@
+package builder
+
+import (
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus/misc"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/google/uuid"
+)
+
+// blockchain is the minimum interface to the blockchain
+// required to build a block
+type blockchain interface {
+	core.ChainContext
+
+	// Header returns the current tip of the chain
+	CurrentHeader() *types.Header
+
+	// StateAt returns the state at the given root
+	StateAt(root common.Hash) (*state.StateDB, error)
+
+	// Config returns the chain config
+	Config() *params.ChainConfig
+}
+
+type Config struct {
+	GasCeil            uint64
+	SessionIdleTimeout time.Duration
+}
+
+type SessionManager struct {
+	sessions      map[string]*builder
+	sessionTimers map[string]*time.Timer
+	sessionsLock  sync.RWMutex
+	blockchain    blockchain
+	config        *Config
+}
+
+func NewSessionManager(blockchain blockchain, config *Config) *SessionManager {
+	if config.GasCeil == 0 {
+		config.GasCeil = 1000000000000000000
+	}
+	if config.SessionIdleTimeout == 0 {
+		config.SessionIdleTimeout = 5 * time.Second
+	}
+
+	s := &SessionManager{
+		sessions:      make(map[string]*builder),
+		sessionTimers: make(map[string]*time.Timer),
+		blockchain:    blockchain,
+		config:        config,
+	}
+	return s
+}
+
+// NewSession creates a new builder session and returns the session id
+func (s *SessionManager) NewSession() (string, error) {
+	s.sessionsLock.Lock()
+	defer s.sessionsLock.Unlock()
+
+	parent := s.blockchain.CurrentHeader()
+
+	chainConfig := s.blockchain.Config()
+
+	header := &types.Header{
+		ParentHash: parent.Hash(),
+		Number:     new(big.Int).Add(parent.Number, common.Big1),
+		GasLimit:   core.CalcGasLimit(parent.GasLimit, s.config.GasCeil),
+		Time:       1000,             // TODO: fix this
+		Coinbase:   common.Address{}, // TODO: fix this
```

Comment:
hmm yeah interesting to think what you would set this to. Two ideas come to mind:
- set to '0x + keccak256("SUAVE")` so its consistent for users 
- allow for the ability to set it. I think there may be some scenarios where you if your a builder who also has a contract controlled by your address that does something different based on profit in a block, then you would want to be able to set this. Happy to spell this out more but I think it's desirable.

Reply:
Agree, I think some of these params should be configurable when you start a session like the `extra` field.